### ISSUE TITLE
Warn if uploading configuration will overwrite someone elses changes

### ISSUE THREAD

**Issue by garethbowen** [medic/medic-conf#113]
Technical users use medic-conf to make configuration changes to instances and hopefully remember to commit the changes to git to track, back up, and share their work.

This can overwrite other configuration changes if...

1. a user has made changes through the admin app,
2. a user has made changes directly to the database,
3. a user made changes with medic-conf but neglected to commit their changes, or
4. a user forgets to pull updates from the repo before making their changes.

In any of these cases the configuration will be overwritten.

Instead medic-conf should warn when the configuration is different from when it was last executed. The user should be given the option to overwrite the changes (`force`) or update their configuration to be in sync (eg: `git stash`, export configuration, `git stash pop`).

**Comment by garethbowen** [comment]
Ready for AT in `113-warn-config-overwrite`

**Comment by newtewt** [comment]
@wtekeu  I'm getting document update conflicts on app-settings when running the same command a second time. 

Starting with a fresh webapp. 
I run `node /home/newtewt/dev/medic-conf/src/bin/medic-conf.js --accept-self-signed-certs --url=https://<user>:<pass>@localhost`
Confirm `y` to all the prompts about not knowing if we are going to overwrite. 
Once it completes, run the same command. The error below is shown. 

```
INFO Starting action: upload-app-settings… 
{ StatusCodeError: 409 - "{\"error\":\"conflict\",\"reason\":\"Document update conflict.\"}\n"
    at new StatusCodeError (/home/newtewt/dev/medic-conf/node_modules/request-promise-core/lib/errors.js:32:15)
    at Request.plumbing.callback (/home/newtewt/dev/medic-conf/node_modules/request-promise-core/lib/plumbing.js:104:33)
    at Request.RP$callback [as _callback] (/home/newtewt/dev/medic-conf/node_modules/request-promise-core/lib/plumbing.js:46:31)
    at Request.self.callback (/home/newtewt/dev/medic-conf/node_modules/request/request.js:185:22)
    at emitTwo (events.js:126:13)
    at Request.emit (events.js:214:7)
    at Request.<anonymous> (/home/newtewt/dev/medic-conf/node_modules/request/request.js:1161:10)
    at emitOne (events.js:116:13)
    at Request.emit (events.js:211:7)
    at IncomingMessage.<anonymous> (/home/newtewt/dev/medic-conf/node_modules/request/request.js:1083:12)
  name: 'StatusCodeError',
  statusCode: 409,
  message: '409 - "{\\"error\\":\\"conflict\\",\\"reason\\":\\"Document update conflict.\\"}\\n"',
  error: '{"error":"conflict","reason":"Document update conflict."}\n',
  options: 
   { method: 'PUT',
     url: 'https://medic:8b30caf11dbd888b@localhost/medic/api/v1/settings?replace=1',
     headers: { 'Content-Type': 'application/json' },
     body: '{\n  "locales": [\n    {\n      "code": "en",\n      "name": "English"\n    },\n    {\n   
```

**Comment by newtewt** [comment]
Also, should we add a assume yes argument so the user doesn't have to press `y` a bunch when uploading a ton of forms?

**Comment by wtekeu** [comment]
@newtewt I have updated the branch. Also, I am not sure I understand the second comment. It is either you press `y` or `Enter`

**Comment by newtewt** [comment]
Thanks I'll review again.  

As to the second comment. I was thinking similar to how apt-get has an assume yes option. So anytime a user would be prompted to select yes/ no it would just automatically answer yes. Not really apart of this ticket but might be a nice to have. 

http://manpages.ubuntu.com/manpages/cosmic/man8/apt-get.8.html

**Comment by newtewt** [comment]
@garethbowen  @SCdF There is an issue where the API updates app_settings after they've been uploaded. This means that if you were to run an upload from medic-conf 2 times in a row with no changes you'd be notified every time that you're going to overwrite changes because the rev has changed in couch. @wtekeu says this is because API doesn't delete existing fields it only overwrites existing and then saves. This seems to only occur if you're updating app_settings. Should we adjust API and how it saves or drop the app_settings check or something else?

**Comment by garethbowen** [comment]
Unfortunately we can't change that API without bumping the API version because other scripts rely on it operating the way it does. We could introduce a new option to make it overwrite completely which would solve this issue but it wouldn't be backwards compatible so old APIs would still be affected.

I wouldn't want to drop the rev check on the app_settings either because it's probably the most likely part to be updated outside of medic-conf (eg: through the admin UI).

@wtekeu How about we adjust the API to return the new rev. Then after it's called we can update the locally stored rev with the latest version?

**Comment by wtekeu** [comment]
@garethbowen It looks like the main issue is that the api listens to settings changes and then compare it the default settings (`config.default.json`) and then performs an update (which changes the rev one more time): https://github.com/medic/medic/blob/master/api/src/config.js#L59-L75

Which is why the rev do never match whenever an attempt to re-upload the settings is executed. And since there is a diff, it displays the prompt.

**Comment by garethbowen** [comment]
Ahh of course! That's a really old "feature" from back in the days when the `config.default.json` was actually useful. It is now so far away from any project's actual config that it's probably not helpful at all any more.

In addition, as part of 3.7.0 @kennsippell is working on #5914 which will replace this `config.default.json` with the CHT reference app. We certainly wouldn't want to merge the reference app with whatever settings someone uploads.

Feel free to delete the config merging code in API. There's a chance that the code is relying on some property being defined in settings but if that happens we should fix the code to be null safe rather than updating the settings to have a default property.

**Comment by dianabarsan** [comment]
I think merging `config.default.json` into the app settings was useful when new features with configs were added, including the addition of new permissions. As far as I know, we can't add permissions in the admin UI, so users would be forced to use medic-conf to create the new permissions.

**Comment by garethbowen** [comment]
You're right - new permissions are why we need this... Thanks!

It still feels clunky to me to merge default every time the settings changes or api starts so I'd like to do something to fix it.

Firstly, I think we should reduce the default to _just_ what's needed. I think this is just permissions but it's worth reviewing the file.

Secondly, should it be applied as it is now, or done as a migration when new permissions are added, or applied to permissions when accessed so the settings doc remains unchanged. I think I'm leaning towards doing it in a migration so it happens once and only once on upgrade, and adding a migration now to make sure everyone's caught up with the current default permissions. What do you think @dianabarsan ?

**Comment by dianabarsan** [comment]
In theory, I think it's enough, in practice .. debatable. 
I fear that, even with this warning feature in place, people will still overwrite configuration.
We should remember how many support requests were about people pushing the old format of permissions after such a migration.

**Comment by garethbowen** [comment]
Do you think we should keep the current merging behaviour, perhaps with _just_ the permissions and nothing else?

**Comment by dianabarsan** [comment]
Just the permissions and only when necessary would be perfect! 
In regards to the use of `api/src/config.default.json`, I think it also kind of filled the gap of us lacking a Schema document post 3.0. I know @SCdF is working on bringing that up to speed and re-adding it.

**Comment by wtekeu** [comment]
@garethbowen please note that this will not prevent the additional update as the extra permissions will result in an update (which means new rev)

**Comment by garethbowen** [comment]
@wtekeu Yes. I think the other thing that should change is the default permissions should be applied in the `/api/v1/settings` endpoint controller so it gets modified _before_ saving to the db. This way only one change happens and the rev should be all good. I think we should also keep the change listener around just in case someone updates the settings doc directly (not through the API) but it shouldn't save when there is no change to the doc. Does that make sense?

**Comment by wtekeu** [comment]
👍

**Comment by wtekeu** [comment]
@garethbowen I was reviewing it more carefully and it looks like it is impossible to know whether the change was the consequence of a direct update or that of an api request. Also, you say that `it shouldn't save when there is no change to the doc` but the listener is triggered only when the doc changed. Why do we need to worry about changes that result from a direct update?

**Comment by garethbowen** [comment]
You're right - in the listener we can't tell if it came from a direct request or through the API.

My thinking is that medic-conf (and other API users) will post the JSON to the settings endpoint, the controller will make sure the config contains all the known permissions, and then store it in the db. Then the changes handler will fire, and make sure the config contains the known permissions again. The key here is that the changes handler should check to see if it's actually changed the JSON at all - if it hasn't then it shouldn't write it to the db. Because there was only one rev written the next upload from medic-conf should just work.

When someone updates the doc directly (which is not recommended but possible) then the settings endpoint will be bypassed. Then when the changes handler fires and applies the known permissions it may actually change the value so it might end up writing a second rev. In this case medic-conf will correctly show a warning next time because the settings were edited directly.

We need to worry about direct update just in case someone has decided to delete all the permissions which would result in unpredictable behaviour. Each permission's configuration can be overwritten as needed but not deleted entirely.

**Comment by kennsippell** [comment]
@wtekeu If this medic-conf feature requires 3.7 to function smoothly, will we only enable it when uploading to API > version 3.7?

**Comment by wtekeu** [comment]
@kennsippell This feature can be used without 3.7. The advantage of using 3.7 is that when TLs will upload app settings it will not always show a prompt indicating that the settings has changed when the change is only the result of default settings properties being added to settings by the API change listener.

**Comment by kennsippell** [comment]
@wtekeu But won't the alert be overactive because the change listener always saves, even when there is no change to the doc? Won't that be annoying for users working with API versions < 3.7?

**Comment by wtekeu** [comment]
@kennsippell as long as TLs know why they are getting the prompt, I think it is moderately annoying (they just have to hit a key). But again, I have never worked with TLs so what may not be annoying to me may be severely annoying to them. Do they update app settings that often?

**Comment by wtekeu** [comment]
@newtewt I have pushed an update which prevents the warning logic from being executed when app settings is uploaded. Same branch medic-conf branch `113-warn-config-overwrite`.
There is a `3.7` webapp branch which has not been reviewed yet (also `113-warn-config-overwrite`) which allows TLs to run the warning logic when app settings is uploaded.  My guess is that you will have to AT a second time.

**Comment by garethbowen** [comment]
This is blocked on https://github.com/medic/medic/issues/5914 as the two medic/medic changes are conflicting.

**Comment by wtekeu** [comment]
@newtewt  ready for AT:
medic-conf: `113-warn-config-overwrite`
medic-webapp: `113-warn-config-overwrite`
(don't merge)

**Comment by newtewt** [comment]
@wtekeu  the app settings portion of this is working. The bad news is that I believe that now that we are generating forms server side. Forms are being edited and the rev is different on the server vs local. So every time you upload app forms you're being informed that you are overwriting a change.  This feature seems like a whack-a-mole effort and I'm betting we'll run into this in future versions.

**Comment by garethbowen** [comment]
Sorry, I should've seen this coming...

I think Nick may be right and using the `_rev` may be too brittle for at least some document types. In the case of forms, we don't care about the generated model and html attachments but everything else in the doc should be identical. We could write a custom function for forms which compares the md5 of the xform attachment and the content of the doc (ignoring the rev) to determine if it's equivalent or not. That sounds like a reasonably large piece of work so we may need to consider deferring this feature to 3.8.0 to give us time to think it through carefully.

Alternatively there may be other simpler solutions that I haven't thought of...

**Comment by garethbowen** [comment]
I think this may also fix #3050

**Comment by garethbowen** [comment]
This has been destabalised by other merges so we've decided to put it off to 3.8.0 to give us more time to come up with a good solution.

**Comment by garethbowen** [comment]
Ready for AT in medic-conf `113-warn-config-overwrite`

**Comment by garethbowen** [comment]
This was already partially ATed by Nick so read through the history to see what changed. The big addition from what was previously reviewed is now we correctly detect if you're overwriting enketo forms. Steps for testing the new part:

1. Upload a configuration to the server to get a known state. Check the _rev of the doc.
2. Upload the configuration again. Check the _rev of the doc is unchanged. Check that you weren't prompted to confirm as you weren't making a change.
3. Copy your configuration directory so you have two copies so you can pretend to be two different users.
4. In the first copy make a change to the form and upload it. Check the _rev changes and your modification is seen in the app. Check you weren't prompted to confirm because your original config was consistent with the server's config.
5. In the second copy make a different change to the form and upload it. Check that you _are_ prompted because now you're overwriting someone else's change. Check that the three options (overwrite, cancel, view diff) all work as expected.

**Comment by ngaruko** [comment]
LGTM. Not sure if this warning isn't somehow ambiguous ? 
```We can't determine if you're going to overwrite someone else's changes or not. Would you like to proceed? [y/n]:```

**Comment by wtekeu** [comment]
@ngaruko 
1. I wanted to be consistent with other `medic-conf` errors handling
2. It is expected .. it checks each document being uploaded .. you will have to do that when (1) you have never used this version of `medic-conf` so doing it for the first time OR (2) there is a mismatch between remote and local docs
